### PR TITLE
Add EvalResultsExporter for writing the results of evaluation to a file

### PR DESCRIPTION
**Pull request by jeffcarp**
Previously you could capture the return value of `estimator.train_and_evaluate`,
which was a dictionary of the final evaluation results (like accuracy, AUC).

Since `estimator.train_and_evaluate` no longer returns a value, it looks like
there's no way to get the `eval_results` dictionary directly. This is the
solution I wrote for getting and storing that dictionary. Since the solution
is generalizable, I thought it'd be useful to include it for others to use.

### CODE REVIEW

**Comment by xiejw** [comment]
train_and_evaluate can return something for local mode. But that behavior does not make sense during distributed mode. This is why it does not return anything.

Jeff, may i ask some information. What's your major use case? local or distributed? And how do you deal with the results saved in json later?

martinwicke@, I think several options could be chosen here
1. returns the eval dict for local mode. But this is very confusing behavior for distributed. 
2. adds a tooling to grab the results. e.g., this pull request or reading events file directly. 

If we focus on local mode, option 1 is the easiest solution as it is convenient.

**Comment by jeffcarp** [comment]
Thanks for the review -

> train_and_evaluate can return something for local mode

I missed how to accomplish that - it looks like [the function itself has no return value](https://github.com/tensorflow/tensorflow/blob/398699286064bc0821056209e1c62065f0e00f82/tensorflow/python/estimator/training.py#L256)?

Our use case is:
- We kick off a (local) training job on ML Engine every 24h
- We want to write the evaluation results somewhere
- We have a script that compares the evaluation results between models
- We can use this script manually to verify a trained model is ready for production, and we plan on in the future using this script to automatically promote models as long as their evaluation results are good

**Comment by rmlarsen** [comment]
@xiejw can you respond to @jeffcarp's question?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
This is a fine Exporter. Note that you also get the evaluation output in the event files used by TensorBoard (as a protobuf, not as JSON, but otherwise the same). 

I don't think we would want to maintain this as part of TensorFlow though.

**Comment by jeffcarp** [comment]
@martinwicke got it, thanks for the review!

**Comment by mufusu** [comment]
I am using the API named 'tf.estimator.train_and_evaluate' for distributed training, but i can't see any evaluation results (like accuracy, AUC) in the worker's LOG.

After reading this pull request Q&A, maybe there are two way to print evaluation results:
1. I have tried Jeff's Exporter, but there's also no logs of export_path or eval_results, i'm confusing for where to get the results. Can you help me? the paper i referenced is https://jeff.is/blog/exporting-eval-results-in-tensorflow.html  @jeffcarp 

2. How can i print the event files used by TensorBoard (as a protobuf, not as JSON, but otherwise the same) in the worker's log? is there any demo for me? Thank you very much! @martinwicke

**Comment by jeffcarp** [comment]
@mufusu - @xiejw mentioned that the behavior is different in distributed mode, perhaps those arguments aren't the same in that case. I haven't tested that exporter in distributed mode. Glancing at the [source](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/training.py#L910) I can't see where exactly it differs. You could try running your trainer in local mode to see if it works there?
